numpy/core/src/umath/loops.c.src

```diff
@@ -2618,41 +2618,48 @@ OBJECT_@kind@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUS
 NPY_NO_EXPORT void
 OBJECT_sign(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
 {
-#if defined(NPY_PY3K)
     PyObject *zero = PyLong_FromLong(0);
+
     UNARY_LOOP {
         PyObject *in1 = *(PyObject **)ip1;
         PyObject **out = (PyObject **)op1;
+        PyObject *ret = NULL;
         int v;
-        PyObject *ret;
-        if (PyObject_Cmp(in1 ? in1 : Py_None, zero, &v) == -1) {
-            return;
+
+        if (in1 == NULL) {
+            in1 = Py_None;
+        }
+
+        v = PyObject_RichCompareBool(in1, zero, Py_LT);
+        if (v == 1) {
+            ret = PyLong_FromLong(-1);
+        } else if (v == 0) {
+            v = PyObject_RichCompareBool(in1, zero, Py_GT);
+            if (v == 1) {
+                ret = PyLong_FromLong(1);
+            } else if (v == 0) {
+                v = PyObject_RichCompareBool(in1, zero, Py_EQ);
+                if (v == 1) {
+                    ret = PyLong_FromLong(0);
+                }
+            }
```

Comment:
I think there is a loophole in this logic. Consider the following:

```
>>> a = np.array([np.nan], dtype=object)
>>> a < 0
array([False], dtype=bool)
>>> a == 0
array([False], dtype=bool)
>>> a > 0
array([False], dtype=bool)
``` 

If we have an object that is neither larger, nor smaller, nor equal to `0`, right now we will set the output to `NULL`, which I am not sure is what we really want. Perhaps `None` is the right thing to set it to, but I'll ask on the list for the community wisdom on this one.